lib/experimental/class-wp-webfonts.php

```diff
@@ -268,150 +386,448 @@ public function validate_webfont( $webfont ) {
+				unset( $variation[ $prop ] );
 			}
 		}
 
-		return $webfont;
+		return $variation;
 	}
 
 	/**
-	 * Register a provider.
+	 * Processes the items and dependencies.
 	 *
-	 * @since 6.0.0
+	 * Processes the items passed to it or the queue, and their dependencies.
 	 *
-	 * @param string $provider The provider name.
-	 * @param string $class    The provider class name.
-	 * @return bool True if successfully registered, else false.
+	 * @since X.X.X
+	 *
+	 * @param string|string[]|false $handles Optional. Items to be processed: queue (false),
+	 *                                       single item (string), or multiple items (array of strings).
+	 *                                       Default false.
+	 * @param int|false             $group   Optional. Group level: level (int), no group (false).
+	 *
+	 * @return array|string[] Array of web font handles that have been processed.
+	 *                        An empty array if none were processed.
 	 */
-	public function register_provider( $provider, $class ) {
-		if ( empty( $provider ) || empty( $class ) ) {
-			return false;
+	public function do_items( $handles = false, $group = false ) {
+		$handles = $this->prep_handles_for_printing( $handles );
+		if ( empty( $handles ) ) {
+			return $this->done;
 		}
-		$this->providers[ $provider ] = $class;
-		return true;
+
+		$this->all_deps( $handles );
+		if ( empty( $this->to_do ) ) {
+			return $this->done;
+		}
+
+		$this->to_do_keyed_handles = array_flip( $this->to_do );
+
+		foreach ( $this->get_providers() as $provider_id => $provider ) {
+			// Alert and skip if the provider class does not exist.
+			if ( ! class_exists( $provider['class'] ) ) {
+				/* translators: %s is the provider name. */
+				trigger_error(
+					sprintf(
+						'Class "%s" not found for "%s" web font provider',
+						$provider['class'],
+						$provider_id
+					)
+				);
+				continue;
+			}
+
+			$this->do_item( $provider_id, $group );
+		}
+
+		$this->process_font_families_after_printing( $handles );
+
+		return $this->done;
 	}
 
 	/**
-	 * Generate and enqueue webfonts styles.
+	 * Prepares the given handles for printing.
 	 *
-	 * @since 6.0.0
+	 * @since X.X.X
+	 *
+	 * @param string|string[]|bool $handles Handles to prepare.
+	 * @return array Array of handles.
 	 */
-	public function generate_and_enqueue_styles() {
-		// Generate the styles.
-		$webfonts = $this->get_webfonts_by_provider( $this->get_enqueued_webfonts() );
-		$styles   = $this->generate_styles( $webfonts );
+	private function prep_handles_for_printing( $handles = false ) {
+		if ( false !== $handles ) {
+			$handles = $this->validate_handles( $handles );
+			// Bail out when invalid.
+			if ( null === $handles ) {
+				return array();
+			}
+		}
 
-		// Bail out if there are no styles to enqueue.
-		if ( '' === $styles ) {
-			return;
+		// Use the enqueued queue.
+		if ( empty( $handles ) ) {
+			if ( empty( $this->queue ) ) {
+				trigger_error( 'No web fonts are enqueued for printing' );
```

Comment:
Do we really need to trigger an error here?

Reply:
In a separate issue and PR(s), let's audit and review all of the notices to share to help devs find where their code is incompatible with the API.